File: companion/ZippieCompanionApp/Design/DiagnosticsScreen.swift
import SwiftUI
import ZippieCompanionKit

/// Why this phone is, or is not, reachable - said in one sentence and then in
/// detail.
///
/// The view RENDERS and does not decide. Every judgement (which failure, which
/// tone, whether via-router counts as good) lives in `Diagnostics` in the Kit,
/// where `swift test` reaches it. That split is not decoration: RelayVerdict
/// exists because a sentence computed inside a SwiftUI view shipped wrong and
/// could not be tested.
struct DiagnosticsScreen: View {
    @StateObject private var model: DiagnosticsModel

    init(consoleHost: String?) {
        _model = StateObject(wrappedValue: DiagnosticsModel(consoleHost: consoleHost))
    }

    var body: some View {
        Page {
            header
            rows
            refresh
        }
        .task { await model.measure() }
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: Space.snug) {
            Text(model.diagnostics.headline)
                .font(Kind.display())
                .tracking(-0.8)
                .foregroundStyle(Ink.primary)
                .fixedSize(horizontal: false, vertical: true)

            // The one thing worth saying above everything else. A phone that is
            // reachable ONLY because some router forwards for it is one SSID
            // change from unmanageable, and that is exactly how a managed Pixel
            // went dark for half an hour.
            if case .viaRouter = model.diagnostics.tailnet {
                Note(text: "This phone has no Tailscale of its own. It can reach "
                   + "the MDM only while on this network.")
            }
        }
        .padding(.bottom, Space.base)
    }

    private var rows: some View {
        VStack(alignment: .leading, spacing: 0) {
            SectionHead(title: "Detail")
            ForEach(Array(model.diagnostics.rows().enumerated()), id: \.offset) { i, row in
                if i > 0 { Hairline() }
                VStack(alignment: .leading, spacing: 2) {
                    Readout(label: row.label, value: row.value, tone: tone(row.tone))
                    if let hint = row.hint {
                        Text(hint)
                            .font(Kind.caption())
                            .foregroundStyle(Ink.secondary)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                }
                .padding(.vertical, 2)
            }
        }
    }

    private var refresh: some View {
        ActionButton(title: model.measuring ? "Measuring..." : "Refresh",
                     role: .quiet, enabled: !model.measuring) {
            Task { await model.measure() }
        }
        .padding(.top, Space.base)
    }

    /// The ONLY place tone becomes colour. Kept to a switch so a new tone in
    /// the Kit is a compile error here rather than a silently grey row.
    private func tone(_ t: DiagnosticRow.Tone) -> Color {
        switch t {
        case .good:    return Ink.primary
        case .bad:     return Ink.down
        case .unknown: return Ink.secondary
        case .note:    return Ink.degraded
        }
    }
}

File: companion-android/app/src/main/java/app/zippie/companion/DiagnosticsScreen.kt
package app.zippie.companion

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import app.zippie.companion.design.Ink
import app.zippie.companion.design.Kind
import app.zippie.companion.design.Space

/**
 * Why this phone is, or is not, reachable - one sentence, then the detail.
 *
 * The composable RENDERS and does not decide. Every judgement (which failure,
 * which tone, whether via-router counts as good) lives in [Diagnostics], where
 * the JVM unit tests reach it. The iOS app splits the same way and for the same
 * reason: a sentence computed inside a view shipped wrong once and could not be
 * tested.
 */
@Composable
fun DiagnosticsScreen(
    diagnostics: Diagnostics,
    measuring: Boolean,
    onRefresh: () -> Unit,
    onBack: () -> Unit,
    // PASSED IN, not read from a Context here, so this stays a pure function of
    // its arguments and can be previewed and tested without device-protected
    // storage. Defaulted so existing callers and previews are unaffected.
    bootLog: List<String> = emptyList(),
) {
    Column(
        Modifier
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 20.dp, vertical = 24.dp),
    ) {
        // A VISIBLE WAY OUT, because the system gesture is not one.
        //
        // This screen shipped with `BackHandler` as its ONLY exit. That is a
        // correct handler and still a trap: the Pixel runs gesture navigation,
        // so there is no back BUTTON anywhere on the device, and the whole
        // screen is a vertical scroller - an edge swipe on a phone in a case,
        // or by anyone who has not learned the gesture, does nothing visible.
        // Reported from the device on 2026-08-12 as simply "cant exit
        // diagnostics screen", which is exactly what it looks like from the
        // outside: a screen with no door.
        //
        // Deliberately at the TOP, before the scrolling content: a control
        // placed after the rows is below the fold on a screen whose length
        // depends on how many diagnostics there are, which is the same bug
        // wearing a different hat.
        TextButton(onClick = onBack, contentPadding = PaddingValues(0.dp)) {
            Text("< Back")
        }
        Spacer(Modifier.height(4.dp))
        Text("DIAGNOSTICS", style = Kind.section)
        Spacer(Modifier.height(6.dp))
        Text(diagnostics.headline, style = Kind.display)

        // The one thing worth saying above everything else. A phone reachable
        // ONLY because some router forwards for it is one SSID change from
        // unmanageable - which is exactly how a managed Pixel went dark.
        if (diagnostics.tailnet is TailnetPath.ViaRouter) {
            Spacer(Modifier.height(10.dp))
            Text(
                "This phone has no Tailscale of its own. It can reach the MDM " +
                    "only while on this network.",
                style = Kind.body,
                color = Ink.secondary,
            )
        }

        Spacer(Modifier.height(20.dp))
        diagnostics.rows().forEachIndexed { i, row ->
            if (i > 0) HorizontalDivider()
            DiagnosticRowView(row)
        }

        Spacer(Modifier.height(20.dp))
        OutlinedButton(onClick = onRefresh, enabled = !measuring) {
            Text(if (measuring) "Measuring..." else "Refresh")
        }

        // THE PHONE'S OWN ACCOUNT OF THE LAST BOOT (#186). Surfaced here
        // because the alternative is adb, and the phone this matters most for
        // is one in a car that nobody can plug into. Newest last, matching the
        // file, so the end of the list is the most recent thing that happened.
        if (bootLog.isNotEmpty()) {
            Spacer(Modifier.height(28.dp))
            Text("BOOT LOG", style = Kind.section)
            Spacer(Modifier.height(6.dp))
            Text(
                "Written to storage that survives a reboot and is readable " +
                    "before unlock. logcat is not.",
                style = Kind.body,
                color = Ink.secondary,
            )
            Spacer(Modifier.height(10.dp))
            bootLog.forEach { line ->
                Text(line, style = Kind.body, color = Ink.secondary)
            }
        }
    }
}

@Composable
private fun DiagnosticRowView(row: DiagnosticRow) {
    Column(Modifier.padding(vertical = 10.dp)) {
        Row(Modifier.fillMaxWidth()) {
            Text(row.label, style = Kind.body)
            Spacer(Modifier.weight(1f))
            Text(
                row.value,
                style = Kind.body,
                color = toneColor(row.tone),
            )
        }
        row.hint?.let {
            Spacer(Modifier.height(2.dp))
            Text(
                it,
                style = Kind.caption,
                color = Ink.secondary,
            )
        }
    }
}

/**
 * The ONLY place a tone becomes a colour, and a `when` with no else so a new
 * tone in [Diagnostics] is a compile error here rather than a silently grey row.
 */
@Composable
private fun toneColor(tone: Tone): Color = when (tone) {
    Tone.GOOD -> Ink.primary
    Tone.BAD -> Ink.down
    Tone.UNKNOWN -> Ink.secondary
    Tone.NOTE -> Ink.degraded
}
